frontend/src/metabase/admin/datamodel/containers/MetricApp.jsx

```diff
@@ -6,73 +6,86 @@ import MetabaseAnalytics from "metabase/lib/analytics";
 
 import MetricForm from "./MetricForm.jsx";
 
-import { metricEditSelectors } from "../selectors";
 import * as actions from "../datamodel";
-import { clearRequestState } from "metabase/redux/requests";
-import { fetchTableMetadata } from "metabase/redux/metadata";
 import { getMetadata } from "metabase/selectors/metadata";
+import Metrics from "metabase/entities/metrics";
+import Tables from "metabase/entities/tables";
 
 const mapDispatchToProps = {
   ...actions,
-  fetchTableMetadata,
-  clearRequestState,
+  createMetric: Metrics.actions.create,
   onChangeLocation: push,
 };
 
 const mapStateToProps = (state, props) => ({
-  ...metricEditSelectors(state, props),
   metadata: getMetadata(state, props),
 });
 
-@connect(
-  mapStateToProps,
-  mapDispatchToProps,
-)
-export default class MetricApp extends Component {
-  async componentWillMount() {
-    const { params, location } = this.props;
+@Metrics.load({
+  id: (state, props) => parseInt(props.params.id),
+  wrapped: true,
+})
+@Tables.load({ id: (state, props) => props.metric.table_id, wrapped: true })
+class UpdateMetricForm extends Component {
+  constructor() {
+    super();
+    this.onSubmit = this.onSubmit.bind(this);
+  }
 
-    let tableId;
-    if (params.id) {
-      const metricId = parseInt(params.id);
-      const { payload: metric } = await this.props.getMetric({ metricId });
-      tableId = metric.table_id;
-    } else if (location.query.table) {
-      tableId = parseInt(location.query.table);
-    }
+  async onSubmit(metric) {
+    await this.props.metric.update(metric);
+    MetabaseAnalytics.trackEvent("Data Model", "Metric Updated");
+    const { id: tableId, db_id: databaseId } = this.props.table;
+    this.props.onChangeLocation(
+      `/admin/datamodel/database/${databaseId}/table/${tableId}`,
+    );
+  }
 
-    if (tableId != null) {
-      // TODO Atte Keinänen 6/8/17: Use only global metadata (`fetchTableMetadata`)
-      this.props.loadTableMetadata(tableId);
-      this.props.fetchTableMetadata(tableId);
+  render() {
+    if (this.props.table) {
+      this.props.table.fetchTableMetadata();
```

Comment:
`render()` should **never** have side-effects like loading data. Typically we use `componentDidMount` and `componentDidUpdate`.

Reply:
Ah sorry I meant to call this out in a comment. `componentDidUpdate` is definitely better than here, but I was wondering if there is some way to have this be part of the initial table loading? Ideally, the table would always have the metadata available in this component.

This might be irrelevant given any changes I make based on your comment that we're already loading some data for all tables. That said, I'd still be curious if there's something in place for "after fetch" actions.

Reply:
> That said, I'd still be curious if there's something in place for "after fetch" actions.

If I understand correctly then no.

You can call `table.fetchTableMetadata()` in both `componentDidMount` and `componentDidUpdate` to make sure it loads both at the beginning and if `table` changes for some reason.

It's also a good idea to check that the prop actually changed in `componentDidUpdate` (even though `fetchTableMetadata` won't issue more than one request, it's a good habit)

Something like.

```javascript
componentDidMount() {
  const { table } = this.props;
  if (table) {
    table.fetchTableMetadata()
  }
}
componentDidUpdate(prevProps) {
  const { table } = this.props;
  if (table && table !== prevProps.table) {
    table.fetchTableMetadata()
  }
}
```